File: benchmarks/llm/lmo1_embedding.py
"""LMO-1 — PyTorch's embedding step, to compare against R2's.

Pairs with `crates/r2-train/examples/lmo1_embedding.rs`. Same tokens, same
dim, same vocabularies.

PyTorch's `nn.Embedding` is a GATHER: `aten::embedding` dispatches to
`aten::index_select`, which copies row tok_i of the table into row i of the
output; its backward is `aten::embedding_dense_backward`, a scatter-add
into the touched rows under `at::parallel_for`. Confirmed by dispatch with
torch.profiler, not assumed from the Python name.

R2 computes the same result as a one-hot matmul, so the comparison is
between two algorithms for one operation rather than between two
implementations of one algorithm. The FLOP counts are printed for both so
the difference is visible rather than inferred.

    R2_LMO_TOKENS=2048 R2_LMO_VOCABS=256,8000,32000 \
      python benchmarks/llm/lmo1_embedding.py
"""

import json
import os
import statistics
import time

import torch


def env_list(key, default):
    raw = os.environ.get(key)
    return [int(x) for x in raw.split(",")] if raw else default


def med_us(reps, fn):
    for _ in range(min(reps, 5)):
        fn()
    out = []
    for _ in range(9):
        s = time.perf_counter()
        for _ in range(reps):
            fn()
        out.append((time.perf_counter() - s) / reps * 1e6)
    return statistics.median(out)


def main():
    torch.set_num_threads(6)
    torch.manual_seed(1)
    t = int(os.environ.get("R2_LMO_TOKENS", 2048))
    d = int(os.environ.get("R2_LMO_DIM", 256))
    vocabs = env_list("R2_LMO_VOCABS", [256, 8000, 32000])

    print("LMO-1 — embedding step only, PyTorch's gather")
    print(f"  {t} tokens, dim {d}, microseconds, median of 9")
    print(f"  torch {torch.__version__}, {torch.get_num_threads()} threads\n")
    print(f"{'vocab':>7} {'onehot MB':>12} {'build':>11} {'forward':>11}"
          f" {'fwd+bwd':>11} {'MFLOP fwd':>13}")
    print("-" * 70)

    torch_rows = {}
    for vocab in vocabs:
        reps = max(3, min(200, int(2e9 / (t * vocab * d)) + 1))
        idx = torch.tensor([(i * 7919) % vocab for i in range(t)], dtype=torch.long)
        table = torch.randn(vocab, d)
        g = torch.randn(t, d)

        # PyTorch builds NO one-hot and needs no such buffer. Shown as 0 so
        # the column lines up with R2's, where it is the dominant cost.
        def fwd():
            with torch.no_grad():
                return torch.nn.functional.embedding(idx, table)

        w = table.clone().requires_grad_(True)

        def fwd_bwd():
            if w.grad is not None:
                w.grad = None
            torch.nn.functional.embedding(idx, w).backward(g)

        f_us = med_us(reps, fwd)
        fb_us = med_us(reps, fwd_bwd)
        # A gather does t*d element copies; there is no multiply-add at all.
        print(f"{vocab:>7} {0.0:>12.1f} {0.0:>11.1f} {f_us:>11.1f}"
              f" {fb_us:>11.1f} {0.0:>13.1f}")
        torch_rows[vocab] = (f_us, fb_us)

    print(f"\nPyTorch's gather does t*d = {t*d} element copies for ANY")
    print("vocabulary and zero multiply-adds, so its cost is flat in vocab")
    print("while the one-hot form's grows linearly.")

    verdict(torch_rows, t, d)


def verdict(torch_rows, t, d):
    """Join R2's numbers with PyTorch's and say who is ahead, per row.

    THE POINT OF THE WHOLE FILE. R2 improving against its own past is not
    a result — the only question is whether R2 is at or above PyTorch on
    this stage. Two programs printing two tables leaves that join to a
    human, and it got skipped. It does not get skipped now.

    Reads `lmo1_r2.json`, written by
    `cargo run --release -p r2-train --example lmo1_embedding`.
    """
    try:
        r2 = json.load(open("lmo1_r2.json"))
    except FileNotFoundError:
        print("\nlmo1_r2.json not found — run the R2 half first:")
        print("  cargo run --release -p r2-train --example lmo1_embedding")
        print("without it there is no verdict, only PyTorch's numbers.")
        return
    if (r2["tokens"], r2["dim"]) != (t, d):
        print(f"\nlmo1_r2.json is for {r2['tokens']} tokens x dim {r2['dim']},")
        print(f"this run is {t} x {d}. Re-run the R2 half at the same shape.")
        return

    print("\n" + "=" * 74)
    print("LMO-1 VERDICT — R2 against PyTorch, same shape, same window")
    print("  GA = the gather both sides ship. fwd+bwd is `.backward(g)` on")
    print("  both: R2 seeds the gradient with `backward_from`, not a")
    print("  manufactured scalar loss.")
    print("=" * 74)
    print(f"{'vocab':>7} {'phase':>9} {'R2 us':>10} {'torch us':>10}"
          f" {'ratio':>8}  verdict")
    print("-" * 74)
    worst = 0.0
    for row in r2["rows"]:
        v = row["vocab"]
        if v not in torch_rows:
            continue
        for phase, r2_us, pt_us in (
            ("fwd", row["ga_fwd"], torch_rows[v][0]),
            ("fwd+bwd", row["ga_fb"], torch_rows[v][1]),
        ):
            ratio = r2_us / pt_us if pt_us > 0 else float("inf")
            worst = max(worst, ratio)
            mark = "R2 AHEAD" if ratio <= 1.0 else f"R2 BEHIND {ratio:.2f}x"
            print(f"{v:>7} {phase:>9} {r2_us:>10.1f} {pt_us:>10.1f}"
                  f" {ratio:>7.2f}x  {mark}")
    print("-" * 74)
    if worst <= 1.0:
        print("LMO-1 PASSES: R2 is at or above PyTorch on every row.")
    else:
        print(f"LMO-1 NOT DONE: worst row is {worst:.2f}x behind PyTorch.")
        print("An R2-versus-R2 speedup does not close this. Only a row at")
        print("or under 1.00x does.")


if __name__ == "__main__":
    main()

File: crates/r2-train/examples/lmo1_embedding.rs
//! LMO-1 — the embedding step, R2 against PyTorch's, and nothing else.
//!
//! R2 computes the token embedding as a ONE-HOT MATMUL (`llm.rs`):
//!
//!     onehot[t x vocab]  with a single 1.0 per row
//!     x = onehot @ table          (t x vocab) @ (vocab x d)
//!
//! PyTorch computes it as a GATHER — `aten::embedding` dispatches to
//! `index_select`, which copies row `tok_i` of the table into row `i` of
//! the output, and its backward is a scatter-add into the touched rows.
//!
//! The two produce the SAME numbers. They do wildly different amounts of
//! work, and the difference scales with the VOCABULARY:
//!
//!     one-hot matmul   2 * t * vocab * d  multiply-adds, and a
//!                      t x vocab buffer built and zeroed every forward
//!     gather           t * d              element copies, no buffer
//!
//! At vocab 256 that ratio is 256; at vocab 8000 it is 8000. The
//! vocabulary was 256 until a BPE tokenizer existed, which is why this
//! stage could be ignored — it cannot be now.
//!
//! Timed through the SHIPPING `Tape`, not a reimplementation, so what is
//! measured is what training runs.
//!
//!     R2_LMO_TOKENS=2048 R2_LMO_VOCABS=256,8000,32000 \
//!       cargo run --release -p r2-train --example lmo1_embedding
//!     python benchmarks/llm/lmo1_embedding.py

use r2_autograd::Tape;

fn median(mut v: Vec<f64>) -> f64 {
    v.sort_by(|a, b| a.partial_cmp(b).unwrap());
    v[v.len() / 2]
}

/// Median of 9 batches. A single reading on this machine is an OS stall
/// until proven otherwise.
fn t_us(reps: usize, mut f: impl FnMut()) -> f64 {
    for _ in 0..reps.min(5) { f(); }
    let mut r = Vec::new();
    for _ in 0..9 {
        let s = std::time::Instant::now();
        for _ in 0..reps { f(); }
        r.push(s.elapsed().as_secs_f64() / reps as f64 * 1e6);
    }
    median(r)
}

fn main() {
    let t: usize = std::env::var("R2_LMO_TOKENS").ok()
        .and_then(|v| v.parse().ok()).unwrap_or(2048);
    let d: usize = std::env::var("R2_LMO_DIM").ok()
        .and_then(|v| v.parse().ok()).unwrap_or(256);
    let vocabs: Vec<usize> = std::env::var("R2_LMO_VOCABS").ok()
        .map(|s| s.split(',').filter_map(|x| x.trim().parse().ok()).collect())
        .unwrap_or_else(|| vec![256, 8000, 32000]);

    println!("LMO-1 — embedding step only, R2's one-hot matmul");
    println!("  {t} tokens, dim {d}, microseconds, median of 9\n");
    println!("{:>7} {:>10} {:>8} {:>11} {:>12} {:>10} {:>10} {:>10}",
             "vocab", "onehot MB", "leaf", "OH fwd", "OH fwd+bwd",
             "GA fwd", "GA fwd+bwd", "speedup");
    println!("  OH = the one-hot matmul this replaces; GA = the gather that ships.");
    println!("  The table's leaf copy is measured and SUBTRACTED from all four.");
    println!("{}", "-".repeat(84));

    // (vocab, oh_fwd, oh_fb, ga_fwd, ga_fb, leaf) kept for the JSON that
    // the PyTorch side joins against.
    let mut rows: Vec<(usize, f64, f64, f64, f64, f64)> = Vec::new();
    for &vocab in &vocabs {
        // Total work per rep is O(t*vocab*d), so hold it roughly fixed.
        let reps = ((2e9 / (t as f64 * vocab as f64 * d as f64)).ceil() as usize)
            .clamp(3, 200);
        let table: Vec<f32> = (0..vocab * d).map(|i| (i as f32 * 0.0007).sin()).collect();
        let tokens: Vec<usize> = (0..t).map(|i| (i * 7919) % vocab).collect();
        let g: Vec<f32> = (0..t * d).map(|i| (i as f32 * 0.005).sin()).collect();

        // ── build the one-hot, exactly as `forward_fused` does ─────────
        let build = t_us(reps, || {
            let mut onehot = vec![0.0f32; t * vocab];
            for (i, &tok) in tokens.iter().enumerate() { onehot[i * vocab + tok] = 1.0; }
            std::hint::black_box(onehot.len());
        });

        // ── the table leaf, measured so it can be SUBTRACTED ───────────
        // `tape.leaf(p.clone(), true)` copies the whole table onto the tape
        // — 8 MB at vocab 8000. That is the tape's ownership cost, paid by
        // every parameter block, not the embedding's. Charging it here
        // would inflate this stage and is exactly the attribution error
        // that produced a wrong LMO-1 number once already.
        let mut tl = Tape::new();
        let leaf_only = t_us(reps, || {
            tl = Tape::new();
            std::hint::black_box(tl.leaf(table.clone(), true));
        });

        // ── forward: build + leaf + matmul ─────────────────────────────
        let mut tp = Tape::new();
        let fwd = t_us(reps, || {
            tp = Tape::new();
            let mut onehot = vec![0.0f32; t * vocab];
            for (i, &tok) in tokens.iter().enumerate() { onehot[i * vocab + tok] = 1.0; }
            let oh = tp.leaf(onehot, false);
            let w = tp.leaf(table.clone(), true);
            let x = tp.matmul(oh, w, t, vocab, d);
            std::hint::black_box(tp.value(x).len());
        });

        // ── forward + backward, which is what a training step runs ─────
        let fb = t_us(reps, || {
            let mut tb = Tape::new();
            let mut onehot = vec![0.0f32; t * vocab];
            for (i, &tok) in tokens.iter().enumerate() { onehot[i * vocab + tok] = 1.0; }
            let oh = tb.leaf(onehot, false);
            let w = tb.leaf(table.clone(), true);
            let x = tb.matmul(oh, w, t, vocab, d);
            tb.backward_from(x, &g);
            std::hint::black_box(tb.grad(w).len());
        });

        // ── the GATHER, which is what ships now ────────────────────────
        // NO subtraction here. The table leaf is created ONCE, outside the
        // timer, which is exactly PyTorch's situation: its parameter tensor
        // persists across calls and only the output is allocated per call.
        // Timing `Tape::new() + leaf + embed` and subtracting a separately
        // measured `leaf` is a difference of two large noisy numbers — it
        // printed 658.6 / 69.9 / 3167.2 us for IDENTICAL work across the
        // three vocabularies, which is not a measurement.
        let g_fwd = {
            let mut r = Vec::new();
            for _ in 0..9 {
                let mut tg = Tape::new();
                let w = tg.leaf(table.clone(), true);
                for _ in 0..reps.min(5) { std::hint::black_box(tg.embed(w, &tokens, d)); }
                let s = std::time::Instant::now();
                for _ in 0..reps { std::hint::black_box(tg.embed(w, &tokens, d)); }
                r.push(s.elapsed().as_secs_f64() / reps as f64 * 1e6);
            }
            median(r)
        };
        // `backward_from(x, g)` IS PyTorch's `.backward(g)`. This used to
        // build `mul(x, g)` then `sum_all` to manufacture a scalar for
        // `backward` — an extra leaf, an extra elementwise multiply, a
        // reduction and all three of their backwards, none of which the
        // PyTorch side ran. It measured ~4 ms of a ~6.6 ms reading, so the
        // published fwd+bwd ratio was a comparison of R2's HARNESS against
        // PyTorch's embedding.
        let g_fb = (t_us(reps, || {
            let mut tg = Tape::new();
            let w = tg.leaf(table.clone(), true);
            let x = tg.embed(w, &tokens, d);
            tg.backward_from(x, &g);
            std::hint::black_box(tg.grad(w).len());
        }) - leaf_only).max(0.0);

        let mflop = 2.0 * t as f64 * vocab as f64 * d as f64 / 1e6;
        // Table leaf subtracted from both, so what remains is the
        // embedding: build the one-hot, multiply, and its backward.
        let fwd_net = (fwd - leaf_only).max(0.0);
        let fb_net = (fb - leaf_only).max(0.0);
        println!("{vocab:>7} {:>10.1} {leaf_only:>8.1} {fwd_net:>11.1} {fb_net:>12.1} {g_fwd:>10.1} {g_fb:>10.1} {:>9.0}x", (t * vocab * 4) as f64 / 1e6,
                 if g_fb > 0.0 { fb_net / g_fb } else { 0.0 });
        rows.push((vocab, fwd_net, fb_net, g_fwd, g_fb, leaf_only));
        let _ = (build, mflop);
    }

    // Emit the numbers so the PyTorch side can JOIN them and print ONE
    // table with a verdict per row. Two programs printing two tables
    // leaves the comparison to a human, and the comparison is the whole
    // point: an R2-versus-R2 improvement that is still behind PyTorch is
    // not a result. `benchmarks/llm/lmo1_embedding.py` reads this file.
    let mut js = format!("{{\n  \"tokens\": {t}, \"dim\": {d},\n  \"rows\": [\n");
    for (i, r) in rows.iter().enumerate() {
        if i > 0 { js.push_str(",\n"); }
        js.push_str(&format!(
            "    {{\"vocab\": {}, \"oh_fwd\": {:.1}, \"oh_fb\": {:.1}, \"ga_fwd\": {:.1}, \"ga_fb\": {:.1}, \"leaf\": {:.1}}}",
            r.0, r.1, r.2, r.3, r.4, r.5));
    }
    js.push_str("\n  ]\n}\n");
    match std::fs::write("lmo1_r2.json", js) {
        Ok(()) => println!("\nwrote lmo1_r2.json — run `python benchmarks/llm/lmo1_embedding.py`\n\
                            for the joint R2-vs-PyTorch table and the per-row verdict."),
        Err(e) => eprintln!("could not write lmo1_r2.json: {e}"),
    }

    println!("\nA gather does t*d = {} element copies for ANY vocabulary.", t * d);
    println!("The one-hot form's cost grows linearly with the vocabulary, so");
    println!("moving from a byte tokenizer to BPE multiplies this stage by ~31x.");
}
